repl/src/main/scala/com/cloudera/livy/repl/PythonInterpreter.scala

```diff
@@ -192,5 +242,41 @@ private class PythonInterpreter(process: Process, gatewayServer: GatewayServer,
       parse(line)
     }
   }
+
+  def createWrapper(driver: ReplDriver, msg: BaseProtocol.BypassJobRequest): BypassJobWrapper = {
+    new BypassJobWrapper(driver, msg.id,
+      new BypassPySparkJob(msg.serializedJob, pysparkJobProcessor))
+  }
+
+  def addFile(path: String): Unit = {
+    pysparkJobProcessor.addFile(path)
+  }
+
+  def addPyFile(driver: ReplDriver, conf: SparkConf, path: String): Unit = {
+    val localCopyDir = new File(pysparkJobProcessor.getLocalTmpDirPath)
+    val localCopyFile = driver.copyFileToLocal(localCopyDir, path, SparkContext.getOrCreate(conf))
+    pysparkJobProcessor.addPyFile(localCopyFile.getPath)
+    if (path.endsWith(".jar")) {
+      driver.addLocalFileToClassLoader(localCopyFile)
+    }
+  }
+
+  def readPort(): Unit = {
```

Comment:
I think this would be cleaner if you hooked all this up into `waitUntilReady`. e.g.

```
val READY_REGEX = "READY\\(port=([0-9]+)\\)".r
stdout.readline() match {
  case null => 
  case READY_REGEX(port) => // set up client
  case _ => // recurse
}
```